src/Runtime/Distributions/Automata/Automaton.StateData.cs

```diff
@@ -0,0 +1,152 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace Microsoft.ML.Probabilistic.Distributions.Automata
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Runtime.Serialization;
+
+    using Microsoft.ML.Probabilistic.Distributions;
+    using Microsoft.ML.Probabilistic.Math;
+    using Microsoft.ML.Probabilistic.Serialization;
+    using Microsoft.ML.Probabilistic.Utilities;
+
+    public abstract partial class Automaton<TSequence, TElement, TElementDistribution, TSequenceManipulator, TThis>
+        where TSequence : class, IEnumerable<TElement>
+        where TElementDistribution : class, IDistribution<TElement>, SettableToProduct<TElementDistribution>, SettableToWeightedSumExact<TElementDistribution>, CanGetLogAverageOf<TElementDistribution>, SettableToPartialUniform<TElementDistribution>, new()
+        where TSequenceManipulator : ISequenceManipulator<TSequence, TElement>, new()
+        where TThis : Automaton<TSequence, TElement, TElementDistribution, TSequenceManipulator, TThis>, new()
+    {
+        /// <summary>
+        /// Represents a state of an automaton that is stored in the Automaton.statesData. This is an internal representation
+        /// of the state. <see cref="State"/> struct should be used in public APIs.
+        /// </summary>
+        [Serializable]
+        internal class StateData
+        {
+            /// <summary>
+            /// The default capacity of the <see cref="transitions"/>.
+            /// </summary>
+            private const int DefaultTransitionArrayCapacity = 1;
+
+            /// <summary>
+            /// The array of outgoing transitions.
+            /// </summary>
+            /// <remarks>
+            /// We don't use <see cref="List{T}"/> here for performance reasons.
+            /// </remarks>
+            [DataMember]
+            private Transition[] transitions = new Transition[DefaultTransitionArrayCapacity];
+
+            /// <summary>
+            /// The number of outgoing transitions from the state.
+            /// </summary>
+            [DataMember]
+            private int transitionCount;
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="StateData"/> class.
+            /// </summary>
+            public StateData()
+            {
+                this.EndWeight = Weight.Zero;
+            }
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="StateData"/> class.
+            /// </summary>
+            /// <param name="transitions">The outgoing transitions.</param>
+            /// <param name="endWeight">The ending weight of the state.</param>
+            [Construction("GetTransitions", "EndWeight")]
+            public StateData(IEnumerable<Transition> transitions, Weight endWeight)
```

Comment:
Would it be horrible to make this an `IReadOnlyCollection<Transition>`? Because you would know the number of states and could allocate the correct number of transitions. And since you're passing an `IEnumerable<Transition>`, you probably know all the transitions already at the call site. If it breaks tons of call sites, at least make a second ctor and make sure to use it where possible.

Reply:
This code is carried from previous version. I would keep it like this for now, because I have bigger transitions overhaul in mind.